sale_blanket_order/models/blanket_orders.py

```diff
@@ -0,0 +1,331 @@
+        if not currency_id:
+            currency_id = product_currency
+            cur_factor = 1.0
+        else:
+            if currency_id.id == product_currency.id:
+                cur_factor = 1.0
+            else:
+                cur_factor = currency_id._get_conversion_rate(
+                    product_currency, currency_id)
+
+        product_uom = product.uom_id.id
+        if uom and uom.id != product_uom:
+            # the unit price is in a different uom
+            uom_factor = uom._compute_price(1.0, product.uom_id)
+        else:
+            uom_factor = 1.0
+
+        return product[field_name] * uom_factor * cur_factor, currency_id.id
+
+    @api.multi
+    def _get_display_price(self, product):
+        # Copied and adapted from the sale module
+        self.ensure_one()
+        pricelist = self.order_id.pricelist_id
+        partner = self.order_id.partner_id
+        if self.order_id.pricelist_id.discount_policy == 'with_discount':
+            return product.with_context(pricelist=pricelist.id).price
+        final_price, rule_id = pricelist.get_product_price_rule(
+            self.product_id, self.original_qty or 1.0, partner)
+        context_partner = dict(self.env.context, partner_id=partner.id,
+                               date=fields.Date.today())
+        base_price, currency_id = self.with_context(context_partner)\
+            ._get_real_price_currency(
+                self.product_id, rule_id, self.original_qty, self.product_uom,
+                pricelist.id)
+        if currency_id != pricelist.currency_id.id:
+            currency = self.env['res.currency'].browse(currency_id)
+            base_price = currency.with_context(context_partner).compute(
+                base_price, pricelist.currency_id)
+        # negative discounts (= surcharge) are included in the display price
+        return max(base_price, final_price)
+
+    @api.multi
+    @api.onchange('product_id', 'original_qty')
+    def onchange_product(self):
+        if self.product_id:
+            self.product_uom = self.product_id.uom_id.id
+            if self.order_id.pricelist_id and self.order_id.partner_id:
+                self.price_unit = self._get_display_price(self.product_id)
+
+    @api.multi
+    @api.depends(
+        'sale_order_lines_ids.blanket_line_id',
+        'sale_order_lines_ids.product_uom_qty',
+        'sale_order_lines_ids.qty_delivered',
+        'sale_order_lines_ids.qty_invoiced',
+        'original_qty'
+    )
+    def _compute_quantities(self):
+        for line in self:
+            sale_lines = line.sale_order_lines_ids
+            line.ordered_qty = sum(l.product_uom_qty for l in sale_lines)
+            line.invoiced_qty = sum(l.qty_invoiced for l in sale_lines)
+            line.delivered_qty = sum(l.qty_delivered for l in sale_lines)
+            line.remaining_qty = line.original_qty - line.ordered_qty
+
+    @api.multi
+    def _validate(self):
```

Comment:
Same here, could this be perhaps replaced by `api.constrains()` and conditions instead of assert?

Reply:
I'll modify asserts but this cannot be constrains because it's called on order validation (action)

Reply:
And finally I think I will let asserts because I like the way it was done (catching AssertionError and raise an Odoo Exception)